### PR TITLE
[PyCDE] Set the module name to differentiate modules with different parameters

### PR DESCRIPTION
**Pull request by teqdruid**

### CODE REVIEW

File: frontends/PyCDE/src/pycde/module.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 #  See https://llvm.org/LICENSE.txt for license information.
 #  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-from __future__ import annotations
+from .hacks import attribute_to_var
```
Comment (mikeurbach) [inline frontends/PyCDE/src/pycde/module.py:5]: This seems to include code from a previous revision.

File: frontends/PyCDE/test/polynomial.py [hunk 2]
```diff
@@ -21,6 +21,8 @@ def __init__(self, name: str, coefficients: list[int], **kwargs):
     """coefficients is in 'd' -> 'a' order."""
     self.instanceName = name
     self.coefficients = Parameter(coefficients)
+    self.module_name = Parameter("PolyComputeForCoeff_" +
```
Comment (mikeurbach) [inline frontends/PyCDE/test/polynomial.py:24]: Having the user specify a magic parameter named `module_name` seems... magic. None of the other parameters get any special treatment. Is the point of this so the generated modules have non-conflicting symbol names? If so, perhaps the __Module constructor can have some generic logic to stringify and concatenate Parameter values. If the point is for the user to directly control the module name for readability, etc., I think we could expose it a different way than a magic Parameter.

File: frontends/PyCDE/test/polynomial.py [hunk 3]
```diff
@@ -71,6 +73,7 @@ def build(self, top):
     x = hw.ConstantOp.create(i32, 23)
     poly = PolynomialCompute("example", [62, 42, 6], x=x)
     PolynomialCompute("example2", [62, 42, 6], x=poly.y)
+    PolynomialCompute("example2", [1, 2, 3, 4, 5], x=poly.y)
```
Comment (mikeurbach) [inline frontends/PyCDE/test/polynomial.py:76]: Add a new `CHECK` which tests this functionality?

File: frontends/PyCDE/src/pycde/module.py (hunk 1 above)
Comment (teqdruid) [reply frontends/PyCDE/src/pycde/module.py]: Yeah, I fucked up a rebase really bad.

File: frontends/PyCDE/test/polynomial.py (hunk 3 above)
Comment (teqdruid) [reply frontends/PyCDE/test/polynomial.py:76]: And I didn't realize that the test was broken. This is what happens when meetings only allow me to work 15 to 30 mins at a time.

File: frontends/PyCDE/test/polynomial.py (hunk 2 above)
Comment (teqdruid) [reply frontends/PyCDE/test/polynomial.py:24]: Yeah, I was concerned about this as well.

Comment (teqdruid) [reply frontends/PyCDE/test/polynomial.py:24]: The intention is to solve both problems. I considered trying to automatically string-ify the parameters, but I've not had much success in the past getting anything readable with that method.

Comment (teqdruid) [reply frontends/PyCDE/test/polynomial.py]: What do you think of the simple method call I switched to? It's more explicit with the same result.

**Comment by teqdruid** [comment]
I realized that this wasn't ready to go until after I created the PR, but couldn't find the button to mark it as a draft.

File: frontends/PyCDE/test/polynomial.py (hunk 2 above)
Comment (mikeurbach) [reply frontends/PyCDE/test/polynomial.py]: The explicit public API to set a module name makes sense, but could that just store a string on `self.module_name`? Seems like it still uses a Parameter under the hood, which isn't the end of the world, but it's still a little confusing that we need to remember to `pop` out that Parameter, if it exists.

Comment (mikeurbach) [reply frontends/PyCDE/test/polynomial.py]: > generic logic to stringify and concatenate Parameter values

FYI I was curious about this, and I took a look. I'll have a PR in a second... the output is fairly readable IMO.

Comment (mikeurbach) [reply frontends/PyCDE/test/polynomial.py]: I drafted my thoughts here: https://github.com/llvm/circt/pull/1230. Let me know what you think.

Comment (teqdruid) [reply frontends/PyCDE/test/polynomial.py]: In some sense it is a design parameter in that is should be available to the generator and needs to be unique'd on. That it requires a parameter on the generate method is unfortunate....

**Review by mikeurbach — APPROVED** [review]

**Comment by teqdruid** [comment]
I'm now thinking I'll wait on this... I don't like the idea of requiring the module name to be an arg in generators.

**Pull request merged**